File: app/src/main/java/com/markdrago/boatzee/domain/DiceState.java
package com.markdrago.boatzee.domain;

import com.markdrago.boatzee.GameConstants;

import java.util.ArrayList;
import java.util.List;

public class DiceState {
    private List<Integer> diceList;
    private List<Boolean> frozenDice;

    public DiceState(List<Integer> diceList) {
        this.diceList = diceList;

        this.frozenDice = new ArrayList<>();
        for (int i = 0; i < GameConstants.DICE_COUNT; i++) {
            this.frozenDice.add(false);
        }
    }

    public DiceState(List<Integer> diceList, List<Boolean> frozenDice) {
        this.diceList = diceList;
        this.frozenDice = frozenDice;
    }

    public Integer nonFrozenDiceCount() {
        int count = 0;
        for (int i = 0; i < GameConstants.DICE_COUNT; i++) {
            if (frozenDice.get(i)) {
                count++;
            }
        }
        return count;
    }

    public DiceState toggleFrozenDie(int position) {
        boolean newState = ! frozenDice.get(position);

        List<Boolean> newFrozenDice = new ArrayList<>(frozenDice);
        newFrozenDice.set(position, newState);
        return new DiceState(diceList, newFrozenDice);
    }
}

File: app/src/main/java/com/markdrago/boatzee/domain/DiceState.kt
package com.markdrago.boatzee.domain

import com.markdrago.boatzee.GameConstants

class DiceState(val nextRoll: () -> Int, val diceList: IntArray, val frozenDice: BooleanArray) {

//    constructor(nextRoll: () -> Int, diceList: IntArray) :
//        this(nextRoll, diceList, BooleanArray(GameConstants.DICE_COUNT, {false}))

    constructor(nextRoll: () -> Int):
        this(nextRoll, IntArray(GameConstants.DICE_COUNT, {x -> nextRoll()}), BooleanArray(GameConstants.DICE_COUNT, {false}))

    fun toggleFrozenDie(pos: Int): DiceState {
        val newFrozenDice = frozenDice.mapIndexed { i, b -> if (i == pos) !b else b}
        return DiceState(nextRoll, diceList, newFrozenDice.toBooleanArray())
    }

    fun partialRoll(): DiceState {
        val newDice = diceList.mapIndexed { i, x ->
            if(frozenDice[i])
                diceList[i]
            else
                nextRoll()
        }

        return DiceState(nextRoll, newDice.toIntArray(), frozenDice)
    }

    override fun toString(): String {
        return diceList.joinToString()
    }
}
